Assigment 1

I want you to use 3 different feature scale methods separately. write a function for this and return the X_scaled variable. Compare the mean squared error results from the 3 methods and report your results. It would be better if you change the code completely into functions (function for cost calculation etc.).


In [33]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler, MinMaxScaler, RobustScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, mean_squared_error
import matplotlib.pyplot as plt
import seaborn as sns


# Generate synthetic data with different feature scales
np.random.seed(0)

# Daily temperature (feature x1) with a range of [20, 40] degrees Celsius
temperature = 20 + 20 * np.random.rand(100, 1)

# Ice cream sales (feature x2) with a range of [0, 200] units
ice_cream_sales = 200 * np.random.rand(100, 1)

# Combine temperature and ice cream sales into a single feature matrix X
X = np.hstack((temperature, ice_cream_sales))

# Daily profit (target variable y)
# Assume that daily profit depends on the temperature, ice cream sales,
# and some random noise (normal distribution)
daily_profit = 1000 + 30 * temperature + 5 * ice_cream_sales + np.random.randn(100, 1)

# Now, you have synthetic data to analyze and build a predictive model for daily profit based on temperature and ice cream sales.

# Perform mean normalization (standardization) on input features

#!!! create a function and scale our X and return X_scaled
#!!! do this for whole method like max, mean and z-score 
#!!! compare the result

def scale_data(X, scaling_technique='standardization'):
    scaling_methods = {
        'standardization': StandardScaler(),
        'min_max': MinMaxScaler(),
        'robust': RobustScaler()
    }

    scaler = scaling_methods.get(scaling_technique)

    if scaler is None:
        raise ValueError("Invalid scaling technique. Choose from 'standardization', 'min_max', or 'robust'.")

    X_scaled = scaler.fit_transform(X)
    return X_scaled


def evaluation_results(y_test, y_pred, w, b):
    mse = mean_squared_error(y_test, y_pred)
    print("Mean Squared Error:", mse)

    # Print the learned coefficients
    print("Learned Coefficients (Weights):")
    print(w)
    print("Bias (Intercept):", b)



In [34]:
# Scale the input features X using mean (robust) normalization
X_scaled = scale_data(X, scaling_technique='robust')

# Define the target variable (daily profit) as the output y
y = daily_profit


# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=0)

# Rest of your code for linear regression, gradient descent, and evaluation
# Initialize the weights and bias terms closer to zero
n_features = X_train.shape[1]
w = np.zeros((n_features, 1)) # Initialize weights with zeros
b = 0

# Use a smaller learning rate for smoother convergence
alpha = 0.01  # Learning rate
num_iterations = 1000

# Implement gradient descent
m = len(X_train)  # Number of training data points

for iteration in range(num_iterations):
    # Compute predictions
    y_pred = np.dot(X_train, w) + b

    # Compute gradients
    w_gradient = (1 / m) * np.dot(X_train.T, (y_pred - y_train))
    b_gradient = (1 / m) * np.sum(y_pred - y_train)

    # Update weights and bias
    w -= alpha * w_gradient.reshape(-1, 1)  # Reshape to match the shape of w
    b -= alpha * b_gradient

# Make predictions on the test data
y_pred = np.dot(X_test, w) + b


# Evaluate the model using Mean Squared Error (MSE)
evaluation_results(y_test, y_pred, w, b)


Mean Squared Error: 195.43204059028065
Learned Coefficients (Weights):
[[277.53635774]
 [416.28767536]]
Bias (Intercept): 2453.8580682040206


In [35]:
print(w)

[[277.53635774]
 [416.28767536]]


In [36]:
# Scale the input features X using z-score normalization (standardization)
X_scaled = scale_data(X, scaling_technique='standardization')

# Define the target variable (daily profit) as a vector y
y = daily_profit


# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# Rest of your code for linear regression, gradient descent, and evaluation
# Initialize the weights and bias terms closer to zero
n_features = X_train.shape[1]
w = np.zeros((n_features, 1))  # Match the shape of the weights to (n_features, 1)
b = 0

# Use a smaller learning rate for smoother convergence
alpha = 0.01  # Learning rate
num_iterations = 1000

# Implement gradient descent
m = len(X_train)  # Number of training data points

for iteration in range(num_iterations):
    # Compute predictions
    y_pred = np.dot(X_train, w) + b

    # Compute gradients
    w_gradient = (1 / m) * np.dot(X_train.T, (y_pred - y_train))
    b_gradient = (1 / m) * np.sum(y_pred - y_train)

    # Update weights and bias
    w -= alpha * w_gradient.reshape(-1, 1)  # Reshape to match the shape of w
    b -= alpha * b_gradient

# Make predictions on the test data
y_pred = np.dot(X_test, w) + b


# Evaluate the model using Mean Squared Error (MSE)
evaluation_results(y_test, y_pred, w, b)

Mean Squared Error: 0.8110670689820058
Learned Coefficients (Weights):
[[172.75879895]
 [276.64937111]]
Bias (Intercept): 2411.492747754207


In [37]:
# Scale the input features X using min-max normalization
X_scaled = scale_data(X, scaling_technique='min_max')

# Define the target variable (daily profit) as a vector y
y = daily_profit


# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# Rest of your code for linear regression, gradient descent, and evaluation
# Initialize the weights and bias terms closer to zero
n_features = X_train.shape[1]
w = np.zeros((n_features, 1))  # Match the shape of the weights to (n_features, 1)
b = 0

# Use a smaller learning rate for smoother convergence
alpha = 0.01  # Learning rate
num_iterations = 1000

# Implement gradient descent
m = len(X_train)  # Number of training data points

for iteration in range(num_iterations):
    # Compute predictions
    y_pred = np.dot(X_train, w) + b

    # Compute gradients
    w_gradient = (1 / m) * np.dot(X_train.T, (y_pred - y_train))
    b_gradient = (1 / m) * np.sum(y_pred - y_train)

    # Update weights and bias
    w -= alpha * w_gradient.reshape(-1, 1)  # Reshape to match the shape of w
    b -= alpha * b_gradient

# Make predictions on the test data
y_pred = np.dot(X_test, w) + b


# Evaluate the model using Mean Squared Error (MSE)
evaluation_results(y_test, y_pred, w, b)

Mean Squared Error: 461.8905818934942
Learned Coefficients (Weights):
[[664.90298743]
 [947.21978952]]
Bias (Intercept): 1599.9096765117727


## Conclusion:

According to the mse results, the best method is the z-score normalization (standardization) method. The mse results are as follows:

- Mean Normalization: 195.43
- Standard: 0.81
- Min-Max: 461.89